cpp/src/merge/merge.cu

```diff
@@ -45,8 +53,57 @@
 
 namespace cudf {
 namespace detail {
+
 namespace {
 
+template <bool has_nulls>
+struct row_lexicographic_tagged_comparator {
+  row_lexicographic_tagged_comparator(table_device_view lhs,
+                                      table_device_view rhs,
+                                      device_span<order const> column_order,
+                                      device_span<null_order const> null_precedence)
+    : _lhs{lhs}, _rhs{rhs}, _column_order{column_order}, _null_precedence{null_precedence}
+  {
+    // Add check for types to be the same.
+    CUDF_EXPECTS(_lhs.num_columns() == _rhs.num_columns(), "Mismatched number of columns.");
+  }
+
+  __device__ bool operator()(index_type lhs_tagged_index,
+                             index_type rhs_tagged_index) const noexcept
+  {
+    auto const [l_side, l_indx] = lhs_tagged_index;
+    auto const [r_side, r_indx] = rhs_tagged_index;
+
+    // Not sure why `const_cast` is needed here
```

Comment:
Does this throw a compiler error without the `const_cast`? Do we need to mark this as `TODO`?

Reply:
What if you change the member `table_device_view _lhs` to be `const`, and change `ptr_[left|right]_dview` to be `const`, too? Pretty much everything in the constructor, members, etc. can be `const`.

Reply:
That worked, although I had no idea why it didn't work previously without the `const_cast` (there indeed were compiler errors)